### PR TITLE
feat(query): add TsCardinalities MPP infrastructure

### PR DESCRIPTION
**Pull request by alextheimer**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

Adds `TsCardinalities` logic to `MultiPartitionPlanner` infrastructure.

### CODE REVIEW

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala [hunk 1]
```diff
@@ -429,6 +431,42 @@ class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider
...
+    // This code is nearly identical to materializeMetadataQueryPlan, but it prevents some
+    //   boilerplate code clutter that results when TsCardinalities extends MetadataQueryPlan.
+    //   If this code's maintenance isn't worth some extra stand-in lines of code (i.e. at matchers that
+    //   take MetadataQueryPlan instances), then just extend TsCardinalities from MetadataQueryPlan.
+
+    import filodb.query.exec.TsCardExec.PREFIX_DELIM
+
+    val queryParams = qContext.origQueryParams.asInstanceOf[PromQlQueryParams]
+    val partitions = partitionLocationProvider.getAuthorizedPartitions(
+      TimeRange(0, Long.MaxValue))
```
Comment (alextheimer) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:444]: I need something here that just returns all partitions, right?

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala [hunk 2]
```diff
@@ -83,6 +83,8 @@ class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider
       materializeSubquery(logicalPlan, qContext)
     } else logicalPlan match {
       case mqp: MetadataQueryPlan               => materializeMetadataQueryPlan(mqp, qContext).plans.head
+      // TODO(a_theimer): awk
+      case lp: TsCardinalities                 => materializeTsCardinalities(lp, qContext).plans.head
       case _                                    => walkLogicalPlanTree(logicalPlan, qContext).plans.head
```
Comment (alextheimer) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:88]: The `TsCardinalities` / `MetadataQueryPlan` separation is starting to get a little awkward, especially given the code for `materializeTsCardinalities`. Maybe in a separate PR these should be merged again?

File: coordinator/src/test/scala/filodb.coordinator/queryplanner/MultiPartitionPlannerSpec.scala [hunk 3]
```diff
@@ -670,6 +669,33 @@ class MultiPartitionPlannerSpec extends AnyFunSpec with Matchers with PlanValida
...
+      override def getPartitions(routingKey: Map[String, String], timeRange: TimeRange): List[PartitionAssignment] =
+        partitions(timeRange)
+
+      override def getAuthorizedPartitions(timeRange: TimeRange): List[PartitionAssignment] =
+        partitions(timeRange)
+    }
+
+    val engine = new MultiPartitionPlanner(partitionLocationProvider, localPlanner, "local", dataset, queryConfig)
+    val lp = TsCardinalities(Seq("a", "b"), 3)
+    val promQlQueryParams = PromQlQueryParams("", startSeconds, step, endSeconds, Some("/api/v1/tscard"))
```
Comment (alextheimer) [inline coordinator/src/test/scala/filodb.coordinator/queryplanner/MultiPartitionPlannerSpec.scala:688]: v1? v1.1? v2?

Comment (alextheimer) [reply coordinator/src/test/scala/filodb.coordinator/queryplanner/MultiPartitionPlannerSpec.scala:688]: sticking with v1

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala [hunk 4]
```diff
@@ -438,6 +440,51 @@ class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider
...
+      }
+      if (prefix.size > 1) {
+        builder.append(s""",_ns_="${prefix(1)}"""")
+      }
+      if (prefix.size > 2) {
+        builder.append(s""",__name__="${prefix(2)}"""")
+      }
+      builder.append("}")
+      builder.mkString
+    }
```
Comment (alextheimer) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:458]: This bakes in the ws/ns/metric setup...

Maybe the `TsCardinalities` constructor should take a mapping of labels to values?

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala (hunk 1 above)
Comment (alextheimer) [reply coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala]: resolved

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala (hunk 4 above)
Comment (alextheimer) [reply coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala]: removed

File: query/src/main/scala/filodb/query/LogicalPlan.scala [hunk 5]
```diff
@@ -125,6 +126,10 @@ case class SeriesKeysByFilters(filters: Seq[ColumnFilter],
                                startMs: Long,
                                endMs: Long) extends MetadataQueryPlan
 
+object TsCardinalities {
+  val SHARD_KEY_LABELS = Seq("_ws_", "_ns_", "__name__")
```
Comment (alextheimer) [inline query/src/main/scala/filodb/query/LogicalPlan.scala:130]: This is baking-in the labels/structure of the shard key prefix, but it's necessary to select/order the shard key label-value pairs from a query. Maybe we could add a shard key prefix spec to the config?

**Review by amolnayak311 — APPROVED** [review]
Approved with some comments. LGTM

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala [hunk 6]
```diff
@@ -75,15 +75,16 @@ class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider
...
       (tsdbQueryParams.isInstanceOf[PromQlQueryParams]
         && !qContext.plannerParams.processMultiPartition)) { // Query was part of routing
       localPartitionPlanner.materialize(logicalPlan, qContext)
-    }  else if (LogicalPlan.hasSubqueryWithWindowing(logicalPlan) || logicalPlan.isInstanceOf[TopLevelSubquery]) {
+    } else if (LogicalPlan.hasSubqueryWithWindowing(logicalPlan) || logicalPlan.isInstanceOf[TopLevelSubquery]) {
       materializeSubquery(logicalPlan, qContext)
     } else logicalPlan match {
-      case mqp: MetadataQueryPlan               => materializeMetadataQueryPlan(mqp, qContext).plans.head
-      case _                                    => walkLogicalPlanTree(logicalPlan, qContext).plans.head
+      case mqp: MetadataQueryPlan => materializeMetadataQueryPlan(mqp, qContext).plans.head
```
[the remark this answers is not in the document]
Comment (alextheimer) [reply coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:85]: Done!

File: query/src/main/scala/filodb/query/PromCirceSupport.scala [hunk 7]
```diff
@@ -45,19 +48,22 @@ object PromCirceSupport {
...
-            card <- c.get[Seq[Map[String, String]]]("cardinality")
-          } yield LabelCardinalitySampl(metric, card)
-
-        case None =>
-          if (c.value.isString)
-            for { label <- c.as[String] } yield LabelSampl(label)
-          else
-            for { metadataMap <- c.as[Map[String, String]] } yield MetadataMapSampl(metadataMap)
+      // TODO: Not the best way to find if this is a LabelCardinality/TsCardinalities response, is there a better way?
+      if (c.downField("metric").focus.nonEmpty) {
```
[the remark this answers is not in the document]
Comment (alextheimer) [reply query/src/main/scala/filodb/query/PromCirceSupport.scala]: Updated

File: query/src/main/scala/filodb/query/LogicalPlan.scala [hunk 8]
```diff
@@ -659,6 +669,7 @@ object LogicalPlan {
         case lp: RawChunkMeta          => lp.filters toSet
         case lp: SeriesKeysByFilters   => lp.filters toSet
         case lp: LabelCardinality      => lp.filters.toSet
+        case lp: TsCardinalities        => lp.prefixFilters.toSet
```
[the remark this answers is not in the document]
Comment (alextheimer) [reply query/src/main/scala/filodb/query/LogicalPlan.scala]: Done

**Pull request merged**